File: src/services/abstract_service.ts
import { ethers } from 'ethers';
import { insertTransactionSchema, type InsertTransaction } from '@shared/schema';
import { storage } from '../storage';

export class AbstractService {
  private apiKey: string;
  private baseUrl: string;
  private agentId: string | null;

  constructor() {
    this.apiKey = "5a693905b3aa8160ff2941f0a42058c100ea2af2ba54e53864ee1166fa6810b7";
    this.baseUrl = "https://agent.api.eternalai.org/api/agent";
    this.agentId = null; // Will be set after creating or retrieving agent
  }

  private async createHeaders() {
    return {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json'
    };
  }

  async parseTransactionCommand(command: string): Promise<InsertTransaction | null> {
    try {
      const response = await fetch('https://api.eternalai.org/v1/chat/completions', {
        method: 'POST',
        headers: await this.createHeaders(),
        body: JSON.stringify({
          chain_id: "8453", // Base network
          model: "DeepSeek-R1-Distill-Llama-70B",
          messages: [
            {
              role: "system",
              content: `Parse transaction command into JSON format with fields:
                {
                  "amount": string,
                  "tokenSymbol": string,
                  "toAddress": string
                }
              Example commands:
              - "Send 100 USDC to 0xabc..."
              - "Transfer 0.1 ETH to 0xdef..."
              - "Pay 50 USDC on Abstract to 0xghi..."
              `
            },
            {
              role: "user",
              content: command
            }
          ],
          response_format: { type: "json_object" }
        })
      });

      if (!response.ok) {
        throw new Error(`Error code: ${response.status} - ${await response.text()}`);
      }

      const data = await response.json();
      const parsedCommand = JSON.parse(data.choices[0].message.content);

      const transaction: InsertTransaction = {
        amount: parsedCommand.amount.toString(),
        tokenSymbol: parsedCommand.tokenSymbol,
        toAddress: parsedCommand.toAddress,
        userId: null // Will be set when user authentication is added
      };

      // Validate the transaction data
      return insertTransactionSchema.parse(transaction);

    } catch (error) {
      console.error('Failed to parse transaction command:', error);
      return null;
    }
  }

  async executeTransaction(transaction: InsertTransaction): Promise<{
    success: boolean;
    transactionHash?: string;
    error?: string;
  }> {
    try {
      // Store transaction in database first
      const storedTx = await storage.createTransaction(transaction);

      // Execute the transaction on Abstract
      // This is where we would integrate with the Abstract protocol
      // For now, we'll simulate a successful transaction
      const txHash = ethers.hexlify(ethers.randomBytes(32));

      // Update transaction status
      await storage.updateTransactionStatus(storedTx.id, "completed", txHash);

      return {
        success: true,
        transactionHash: txHash
      };

    } catch (error) {
      console.error('Transaction failed:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  // Add more methods for Abstract protocol integration later
}

export const abstractService = new AbstractService();

File: src/services/abstract_service.py
"""Base class for all services with LLM-friendly interfaces"""
import logging
import os
from typing import Dict, Any, Optional, List, TypeVar, Generic
from datetime import datetime, timezone
import pytz

T = TypeVar('T')

class AbstractService:
    """Base service class providing standardized interfaces for LLM interaction"""

    def __init__(self):
        self.logger = logging.getLogger(self.__class__.__name__)
        # Set timezone to US Central Standard Time
        self.timezone = pytz.timezone('America/Chicago')

    def format_response(
        self,
        data: Optional[T],
        success: bool = True,
        error: Optional[str] = None
    ) -> Dict[str, Any]:
        """Format service response in a consistent way for LLM consumption"""
        current_time = datetime.now(self.timezone)
        return {
            'success': success and data is not None,
            'data': data,
            'error': error,
            'service': self.__class__.__name__,
            'timestamp': current_time.isoformat()
        }

    def log_error(self, error: Exception, context: str = "") -> None:
        """Standardized error logging"""
        current_time = datetime.now(self.timezone)
        self.logger.error(f"[{current_time.isoformat()}] {context}: {str(error)}")

    def validate_params(self, **kwargs) -> tuple[bool, Optional[str]]:
        """Validate input parameters"""
        required_params = getattr(self, 'REQUIRED_PARAMS', [])
        for param in required_params:
            if param not in kwargs or kwargs[param] is None:
                return False, f"Missing required parameter: {param}"
        return True, None

    def get_current_timestamp(self) -> float:
        """Get current timestamp in US Central timezone"""
        return datetime.now(self.timezone).timestamp()

    async def execute_query(
        self,
        query_type: str,
        **kwargs
    ) -> Dict[str, Any]:
        """Execute a query based on type and parameters"""
        valid, error = self.validate_params(**kwargs)
        if not valid:
            return self.format_response(None, False, error)

        try:
            # Map query type to method
            method = getattr(self, f"query_{query_type}", None)
            if not method:
                return self.format_response(
                    None,
                    False,
                    f"Unknown query type: {query_type}"
                )

            result = await method(**kwargs)
            return self.format_response(result)

        except Exception as e:
            self.log_error(e, f"Error executing {query_type}")
            return self.format_response(None, False, str(e))

    async def describe_capabilities(self) -> Dict[str, Any]:
        """Return service capabilities for LLM understanding"""
        return {
            'name': self.__class__.__name__,
            'description': self.__doc__,
            'query_types': [
                method[6:] for method in dir(self)
                if method.startswith('query_')
            ],
            'required_params': getattr(self, 'REQUIRED_PARAMS', []),
            'timezone': str(self.timezone)
        }

import os
import logging
from typing import Dict, Any, Optional
import requests
from src.connections.base_connection import BaseConnection

logger = logging.getLogger("services.abstract")

class AbstractTransactionService(AbstractService):
    """Service for handling Abstract protocol transactions using natural language"""

    def __init__(self):
        super().__init__()
        self.api_key = os.getenv("ABSTRACT_API_KEY")
        if not self.api_key:
            raise ValueError("Missing ABSTRACT_API_KEY in environment variables")

        self.base_url = os.getenv("ABSTRACT_BASE_URL", "https://agent.api.eternalai.org/api/agent")
        self.agent_id = None  # Will be set after creating or retrieving agent
        self.abstract_address = os.getenv("ABSTRACT_ADDRESS")
        self.abstract_private_key = os.getenv("ABSTRACT_PRIVATE_KEY")

        if not all([self.abstract_address, self.abstract_private_key]):
            raise ValueError("Missing Abstract wallet configuration")

        # Initialize headers
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }

        # Setup connection manager
        class ConnectionManager:
            def __init__(self):
                self.connections = {
                    "abstract": BaseConnection({
                        "address": os.getenv("ABSTRACT_ADDRESS"),
                        "private_key": os.getenv("ABSTRACT_PRIVATE_KEY")
                    })
                }

        self.connection_manager = ConnectionManager()

    async def get_agent_info(self, agent_id: str) -> Dict[str, Any]:
        """Get information about an EternalAI agent"""
        try:
            response = requests.get(
                f"{self.base_url}/{agent_id}",
                headers=self.headers
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            self.log_error(e, "Failed to get agent info")
            return self.format_response(None, False, f"Failed to get agent info: {str(e)}")

    async def process_command(self, command: str) -> Dict[str, Any]:
        """Process a natural language transaction command"""
        try:
            # Import locally to avoid circular dependencies
            from src.actions.abstract_nlp import parse_transaction_command, execute_nlp_command
            
            # Parse the natural language command
            parsed = parse_transaction_command(command)

            if not parsed.get("success", True):
                return parsed

            # Execute the transaction
            result = await execute_nlp_command(self, command)
            return result

        except Exception as e:
            self.log_error(e, "Failed to process command")
            return self.format_response(None, False, f"Transaction failed: {str(e)}")

    def validate_address(self, address: str) -> bool:
        """Validate an Ethereum address"""
        try:
            # Basic format validation
            if not address.startswith("0x"):
                return False
            if len(address) != 42:  # 0x + 40 hex chars
                return False
            # Check if it contains only valid hex characters after 0x
            int(address[2:], 16)
            return True
        except ValueError:
            return False

    async def get_token_balance(self, token_symbol: str) -> Dict[str, Any]:
        """Get token balance for the configured wallet"""
        try:
            # Using EternalAI agent API to get balance info
            if self.agent_id:
                response = requests.get(
                    f"{self.base_url}/{self.agent_id}",
                    headers=self.headers
                )
                response.raise_for_status()
                data = response.json()
                return self.format_response(
                    {
                        "balance": data["result"]["wallet_balance"],
                        "symbol": token_symbol
                    }
                )
            return self.format_response({"balance": 0, "symbol": token_symbol})
        except Exception as e:
            self.log_error(e, "Failed to get balance")
            return self.format_response(None, False, f"Balance check failed: {str(e)}")